[quant][tests] Enable tests to run on all qengine backends

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37890 [quant][tests] Update test_quantize_script to test multiple backends
* **#37843 [quant][tests] Enable tests to run on all qengine backends**

Summary:
Refactor tests to use supported_qengines

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21409626](https://our.internmc.facebook.com/intern/diff/D21409626)

supriyar committed:
[quant][tests] Enable tests to run on all qengine backends
Summary:
Refactor tests to use supported_qengines

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on test/quantization/test_backward_compatibility.py:
```diff
@@ -67,76 +68,55 @@ def remove_prefix(text, prefix):
-        " with instruction set support avx2 or newer.",
-    )
+        for i, qengine in enumerate(supported_qengines):
+            with override_quantized_engine(qengine):
```
optional: thoughts on using a decorator for `with override_quantized_engine(qengine)`, so we can save a level of indent for the test cases?  something like below, and it could wrap the "exit if not supported" logic too

```
@override_qengines
def test_foo(...):
    ...
```

supriyar replied:
I did think about it, but wasn't sure how it would play with the hypothesis params we have and if it would make errors harder to reproduce. Let me try it out

supriyar committed:
Update on "[quant][tests] Enable tests to run on all qengine backends"


Summary:
Refactor tests to use supported_qengines

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][tests] Enable tests to run on all qengine backends"


Summary:
Refactor tests to use supported_qengines

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21409626](https://our.internmc.facebook.com/intern/diff/D21409626)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][tests] Enable tests to run on all qengine backends"


Summary:
Refactor tests to use supported_qengines

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21409626](https://our.internmc.facebook.com/intern/diff/D21409626)

[ghstack-poisoned]

mruberry commented:
Unlanding. This was breaking several builds, like pytorch_windows_vs2017_14.11_py36_cpu_test2. Sample log:

```
Tests the correctness of the batchnorm3d module. ... ok (0.061s)
  test_conv2d_api (quantization.test_quantized_module.TestStaticQuantizedModule) ... Falsifying example: test_fn(
    <quantization.test_quantized_module.TestStaticQuantizedModule testMethod=test_conv2d_api>,
    H=4,
    W=4,
    W_scale=[0.2],
    W_zero_point=[0],
    X_scale=1.2,
    X_zero_point=0,
    Y_scale=4.2,
    Y_zero_point=0,
    batch_size=1,
    dilation=1,
    groups=1,
    in_channels_per_group=2,
    kernel_h=1,
    kernel_w=1,
    out_channels_per_group=2,
    pad_h=0,
    pad_w=0,
    stride_h=1,
    stride_w=1,
    use_bias=False,
    use_channelwise=False,
    use_fused=False,
)
ERROR (0.094s)
  test_conv3d_api (quantization.test_quantized_module.TestStaticQuantizedModule) ... ok (0.226s)
  test_layer_norm (quantization.test_quantized_module.TestStaticQuantizedModule)
Tests the correctness of the layernorm module. ... ok (0.006s)
  test_linear_api (quantization.test_quantized_module.TestStaticQuantizedModule) ... Falsifying example: test_fn(
    <quantization.test_quantized_module.TestStaticQuantizedModule testMethod=test_linear_api>,
    batch_size=1,
    in_features=16,
    out_features=4,
    per_channel=False,
    use_bias=False,
    use_fused=False,
)
ERROR (0.014s)
  test_pool_api (quantization.test_quantized_module.TestStaticQuantizedModule)
Tests the correctness of the pool module. ... ok (0.028s)
  test_quant_dequant_api (quantization.test_quantized_module.TestStaticQuantizedModule) ... ok (0.008s)
  test_relu (quantization.test_quantized_module.TestStaticQuantizedModule) ... ok (0.003s)

======================================================================
ERROR [0.056s]: test_linear_api (quantization.test_quantized_module.TestDynamicQuantizedModule)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\circleci\project\test\quantization\test_quantized_module.py", line 35, in test_fn
    for qengine in supported_qengines:
  File "C:\Jenkins\Miniconda3\lib\site-packages\hypothesis\core.py", line 1116, in wrapped_test
    raise the_error_hypothesis_found
  File "C:\Users\circleci\project\test\quantization\test_quantized_module.py", line 38, in test_fn
    return result
UnboundLocalError: local variable 'result' referenced before assignment

======================================================================
ERROR [0.094s]: test_conv2d_api (quantization.test_quantized_module.TestStaticQuantizedModule)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\circleci\project\test\quantization\test_quantized_module.py", line 35, in test_fn
    for qengine in supported_qengines:
  File "C:\Jenkins\Miniconda3\lib\site-packages\hypothesis\core.py", line 1116, in wrapped_test
    raise the_error_hypothesis_found
  File "C:\Users\circleci\project\test\quantization\test_quantized_module.py", line 38, in test_fn
    return result
UnboundLocalError: local variable 'result' referenced before assignment

======================================================================
ERROR [0.014s]: test_linear_api (quantization.test_quantized_module.TestStaticQuantizedModule)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\circleci\project\test\quantization\test_quantized_module.py", line 35, in test_fn
    for qengine in supported_qengines:
  File "C:\Jenkins\Miniconda3\lib\site-packages\hypothesis\core.py", line 1116, in wrapped_test
    raise the_error_hypothesis_found
  File "C:\Users\circleci\project\test\quantization\test_quantized_module.py", line 38, in test_fn
    return result
UnboundLocalError: local variable 'result' referenced before assignment

----------------------------------------------------------------------
Ran 230 tests in 218.093s

FAILED (errors=3, skipped=40)
```
